tuf/builder.go

```diff
@@ -0,0 +1,402 @@
+package tuf
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/notary/certs"
+	"github.com/docker/notary/trustmanager"
+	"github.com/docker/notary/tuf/data"
+	"github.com/docker/notary/tuf/signed"
+)
+
+// ErrBuildDone is returned when any functions are called on RepoBuilder, and it
+// is already finished building
+type ErrBuildDone struct{}
+
+func (e ErrBuildDone) Error() string {
+	return "the builder is done building and cannot accept any more input or produce any more output"
+}
+
+// ErrBuildFailed is returned when any functions are called on RepoBuilder, and it
+// is already failed building and will not accept any other data
+type ErrBuildFailed struct{}
+
+func (e ErrBuildFailed) Error() string {
+	return "the builder has failed building and cannot accept any more input or produce any more output"
+}
+
+// ErrInvalidBuilderInput is returned when RepoBuilder.Load is called
+// with the wrong type of metadata for thes tate that it's in
+type ErrInvalidBuilderInput struct{ msg string }
+
+func (e ErrInvalidBuilderInput) Error() string {
+	return e.msg
+}
+
+// RepoBuilder is an interface for an object which builds a tuf.Repo
+type RepoBuilder interface {
+	Load(roleName string, content []byte, minVersion int) error
+	LoadRoot(content []byte, minVersion int) error
+	LoadSnapshot(content []byte, minVersion int) error
+	LoadTimestamp(content []byte, minVersion int) error
+	LoadTargets(content []byte, minVersion int) error
+	LoadDelegation(role string, content []byte, minVersion int) error
+	Finish() (*Repo, error)
+	BootstrapNewBuilder() RepoBuilder
+	GetRepo() *Repo
+}
+
+// NewRepoBuilder is the only way to get a pre-built RepoBuilder
+func NewRepoBuilder(certStore trustmanager.X509Store, gun string, cs signed.CryptoService,
+	rootRole *data.BaseRole, rootChecksummer data.ChecksumValidator) RepoBuilder {
+
+	return &repoBuilder{
+		repo:                 NewRepo(cs),
+		rootRole:             rootRole,
+		rootChecksummer:      rootChecksummer,
+		gun:                  gun,
+		certStore:            certStore,
+		cs:                   cs,
+		loadedNotChecksummed: make(map[string][]byte),
+	}
+}
+
+type repoBuilder struct {
+	finished bool
+	failed   bool
```

Comment:
documenting IRL discussion with @cyli : I'm not sure if we want these flags to be state on the repoBuilder itself, or if we need separate cases for `failed` and `finished`.  It would also simplify the switch statements in the `RepoBuilder.Load*` functions